tools/src/main/scala/scala/scalanative/build/Discover.scala

```diff
@@ -12,14 +12,94 @@ import scalanative.build.IO.RichPath
  */
 object Discover {
 
-  /** Compilation mode name that takes SCALANATIVE_MODE into account or default otherwise. */
+  /** List of source patterns used */
+  val jarExtension  = ".jar"
+  val srcExtensions = Seq(".c", ".cpp", ".S")
+  val srcPatterns   = srcExtensions.mkString("glob:**{", ",", "}")
+
+  /** Original jar or dir path and generated dir path for native code */
+  case class NativeLib(src: Path, dest: Path)
+
+  /** To positively identify nativelib */
+  val nativeLibMarkerFile = "org_scala-native_nativelib.txt"
+
+  private lazy val dirMarkerFilePattern = "glob:**" + nativeLibMarkerFile
+
+  def isJar(path: Path): Boolean = path.toString().endsWith(jarExtension)
+
+  object NativeLib {
+    def isJar(nativelib: NativeLib): Boolean = Discover.isJar(nativelib.src)
+  }
+
+  /** Compilation mode name from SCALANATIVE_MODE env var or default. */
   def mode(): String =
     getenv("SCALANATIVE_MODE").getOrElse(build.Mode.default.name)
 
-  /** LTO variant used for release mode. */
+  /** LTO variant used for release mode from SCALANATIVE_LTO env var or default. */
   def LTO(): String =
     getenv("SCALANATIVE_LTO").getOrElse("none")
 
+  /** GC variant used from SCALANATIVE_GC env var or default. */
+  def GC(): String =
+    getenv("SCALANATIVE_GC").getOrElse(build.GC.default.name)
+
+  private[build] def findNativeLibs(classpath: Seq[Path],
+                                    workdir: Path): Seq[NativeLib] = {
+    val nativeLibPaths = classpath.flatMap { path =>
+      if (isJar(path)) readJar(path)
+      else readDir(path)
+    }
```

Comment:
Is this early filter even necessary? Later, `unpackNativeCode` will anyway copy only relevant files, ending up being a no-op if there is no relevant file. The early filter has a cost, since we must walk through all the files.

Reply:
This filter reduces the search space down to only the classpath entries that need native processing. This native group then gets named/numbered according to your scheme. This list has to have at least one entry for the `nativelib` otherwise there is a problem. This much smaller list is then searched to find the `nativelib`.

We could also cache this if desired. The only downside would be if you decided to add a C file to a project that was not already in the list you would have to reload `sbt`. This is not too much different that if you add or change a dependency you must restart `sbt`. In this case the dependency would be changing from one that has no native code to one that does. I did not time or benchmark but I can't detect any slow down for searching.

Reply:
We only want `unpackNativeCode` to work on libraries with native code. This blindly unpacks jars and copies directories and hashes the jars and directory contents. Later on in the process the code is found to be compiled. On a previous commit I combined the search into one operation using one glob for each file extension.